concurrency/runner_test.go

```diff
@@ -72,6 +72,99 @@ func TestForEachUser_ShouldReturnImmediatelyOnNoUsersProvided(t *testing.T) {
 	}))
 }
 
+func TestForEachJob(t *testing.T) {
+	var (
+		ctx = context.Background()
+	)
+
+	jobs := []string{"a", "b", "c"}
+	processed := make([]string, len(jobs))
+
+	err := ForEachJob(ctx, len(jobs), 2, func(ctx context.Context, idx int) error {
+		processed[idx] = jobs[idx]
+		return nil
+	})
+
+	require.NoError(t, err)
+	assert.ElementsMatch(t, jobs, processed)
+}
+
+func TestForEachJob_ShouldBreakOnFirstError_ContextCancellationHandled(t *testing.T) {
+	var (
+		ctx = context.Background()
+
+		// Keep the processed jobs count.
+		processed atomic.Int32
+	)
+
+	err := ForEachJob(ctx, 3, 2, func(ctx context.Context, idx int) error {
+		if processed.CAS(0, 1) {
+			return errors.New("the first request is failing")
+		}
+
+		// Wait 1s and increase the number of processed jobs, unless the context get canceled earlier.
+		select {
+		case <-time.After(time.Second):
+			processed.Add(1)
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+
+		return nil
+	})
+
+	require.EqualError(t, err, "the first request is failing")
+
+	// Since we expect the first error interrupts the workers, we should only see
+	// 1 job processed (the one which immediately returned error).
+	assert.Equal(t, int32(1), processed.Load())
+}
+
+func TestForEachJob_ShouldBreakOnFirstError_ContextCancellationUnhandled(t *testing.T) {
```

Comment:
What exactly does this test checks as opposed to `TestForEachJob_ShouldBreakOnFirstError_ContextCancellationHandled`?

Reply:
From what I see, `TestForEachJob_ShouldBreakOnFirstError_ContextCancellationHandled` is a simple test that returns an error and makes sure that context is canceled, without extra assertions on how many jobs were launched.


This `TestForEachJob_ShouldBreakOnFirstError_ContextCancellationUnhandled` makes sure that when two jobs (with concurrency=2) are launched, then the third job is not launched at all.